### ISSUE TITLE
Feature suggestion: allow language-dependent inline code or paragraphs.

### ISSUE THREAD

**Issue by hobovsky** [Codewars/codewars.com#1556]
Currently we have language-dependent code blocks, which are useful for putting examples, stubs, etc. in the description. But it's often not sufficient, because descriptions can differ between languages not only in terms of examples.

I have two use cases in mind:

1. Name of variable/class/method differs between languages. Then we get description like: "Complete the function `DoSomething` (`doSomething` in Java and JavaScript, `do_something` in Python, `@ds` in MUMPS) so it returns blah blah blah..." - in this case, only code sections differ between languages.

2. Whole sentence or paragraph needs to be constructed in another way in different languages, like: "Function should return length of the path, or `null` if no path is found (in C and C++ it should return -1 instead of null)".

It would be great if some paragraphs, or parts of them, could change when language is changed from the dropdown. I think it would make translations a bit less cluttered and descriptions would be easier to translate.

**Comment by FArekkusu** [comment]
IMO users should raise an issue whenever they see trash descriptions like:
```
Complete the function DoSomething (doSomething in Java and JavaScript, do_something in Python, @ds in MUMPS) so it returns...
```
As soon as the author starts mentioning all the unnecessary details like function/parameter names, the translators do the same thing (because *reasons*) and we end up with more and more rubbish in the description.

Although, having inline language-dependant blocks would be a good feature. For example, if one has to write some language-specific notes in the description it's:
a) visible in all languages
b) mashed up into description-ruining trash like `return None/null/nil in Python/JS/Ruby or -1 in C/C++ etc.`
c) the same note has to be rewritten with different if-blocks to include the same *slightly* different message.

**Comment by Voileexperiments** [comment]
Listing the *implementation details* of every single language is often a red flag that implies these thing might have happened:

- The kata author somehow thinks that the users who attempt the kata are `8kyu` newbies
- There are no sample tests/the actual tests are the same as the descriptions
- The translations might have been written with a different spec
- The kata author submitted to a newbie's bogus issue in the comments saying "but my language doesn't have `null`/something! You should put that in the descriptions"

In other words, it's bad, and it distracts people from the actual descriptions that describes the task. Also, it's the sole reason for all the translation description conflict sinkholes, so it's definitely not recommended.

**Comment by hobovsky** [comment]
> As soon as the author starts mentioning all the unnecessary details like function/parameter names, the translators do the same thing (because reasons) and we end up with more and more rubbish in the description.

> Listing the implementation details of every single language is often a red flag that implies these thing might have happened: [...]

90% agreed, but still:
 - some ideas cannot be translated universally between languages. For example, NASM and BF often require more detailed information on function contract (input/output addresses and registers), passing arrays in C might require additional information, input/output types might have no equivalents between translations (I think I've seen things like "return a resulting _[data structure]_, or a string formatted in following way in _[language which does not support given data structure]_", but I can't recall specific kata/language right now)
 - There are katas like this one: https://www.codewars.com/kata/path-finder-number-2-shortest-path which was initially created in JS (I believe) and requires return value of calculated length, or `null` if not found. Beta process did not catch it, no one thought that mixing ints and nulls might be not the best idea, and when someone wanted to translate the kata into C++ _a year and a half_ after kata was created, they were out of options.

In ideal world, when betas would be moderated really carefully with many aspects in mind, and could be easily changed after accepting, it would be less of a problem. But right now, translators need to work with what they got, and sometimes they got sub-par quality of initial version (or are limited by the language they translate to). Or maybe you know a good way to proceed in such cases?

**Comment by g964** [comment]
OK, weird:-( but written in February 2015. Feel free to modify the description!

**Comment by anter69** [comment]
As already suggested above: clean up the description to get rid of the unnecessary details, and use [optional code blocks](https://github.com/Codewars/codewars.com/wiki/Markdown-Formatting#optional-section-formatting) when really needed.

**Comment by hobovsky** [comment]
As I have an impression that you guys focus on the other side of the problem I have in mind, let me have one more (last) take: I do not mean situations when description can be cleaned up, implementation details can be abstracted away, and brought to the form which would fit all versions. If it could be done, then do it, no problem. What I have in mind are following situations:

 - kata already exists, is fairly mature, but has a design flaw - like the one with "return int or null". As in many languages **int**s are not nullable, description has to diverge into two directions and there is no other possibility (or is there?). Changing the behavior of the kata is also not possible, because translator often does not have such power, and besides it would change the behavior of a mature kata with many solutions. Or I am wrong here on this point? And please don't say that such katas do not exists, they should be fixed in beta or that it should not happen in well authored kata, because we do have such katas that are really old, out of beta, accepted, and with many solutions. What **is** and what **should be** are two different things.

 - some aspects of the kata cannot be translated because of unique language features. For example, I can imagine a kata with following description:

> [...] Return list of calculated results.
> *Note for NASM:* address of output buffer is given in **rdi**.
> *Note for C:* output buffer is preallocated and you do not have to free it
> *Note for shell:* return result as a string of numbers separated by semicolon
> *Note for...*

Conditional code block can be used for such cases, but these are not, well, a code.

Anyway, it's just an idea. If you do not find it useful or worth implementing, it's ok. If you have general ideas how to proceed in such situations, I am interested to hear them.

**Comment by FArekkusu** [comment]
> Conditional code block can be used for such cases, but these are not, well, a code.

This is why there're if-language blocks which allow to write language-specific **text**.

**Comment by anter69** [comment]
@hobovsky wrote:
> Conditional code block can be used for such cases, but these are not, well, a code.

That was just my bad wording: it's not *"optional code block"* but *"optional section formatting"*. And as such, it can hold any kind of markdown text, not only code -- see the example in the documentation. I just tried it, and it works properly on CW.

So your example would be:
```
~~~if:nasm
Note for NASM: address of output buffer is given in rdi.
~~~
~~~if:c
Note for C: output buffer is preallocated and you do not have to free it
~~~
~~~if:shell
Note for shell: return result as a string of numbers separated by semicolon
~~~
```

**Comment by Voileexperiments** [comment]
I think the point is: adding language-specific instructions should be *exceptions*, not the norm. Typically it shouldn't be done (or you get the description conflict sinkholes), and *then* for the other cases it can be avoided by:

- not coming up with JS-oriented kata specs in the first place that other languages have great trouble following
- translators not doing funny things and change the spec silently in their translations
- using types, conventions and *code comments* to do this instead

Actually I think almost all of those cases should go to initial code comments: after all, initial code is free, and comments are for exactly this kind of scenario: explanations for the specific code. Besides, the Haskell katas have been putting lots of text in code comments in a long time, and it works just as fine.

**Comment by hobovsky** [comment]
> This is why there're if-language blocks which allow to write language-specific text.

> That was just my bad wording: it's not "optional code block" but "optional section formatting". And as such, it can hold any kind of markdown text, not only code [...]

And this is EXACTLY what I was asking for! Thanks for introducing this useful feature so quickly! :)
I am a moron and I should RTFM better. Sorry for that guys, and thanks for the info.
